File: scala/src/main/scala/FirstDuplicate.scala
import scala.collection.mutable

/**
 * Implementation A: Using mutable.HashSet.
 *
 * Iterate through the stream and add each value to the hash set.
 * If add returns false (the value was already present), we've found
 * the first duplicate. O(n) expected time, O(n) space.
 *
 * mutable.HashSet.add returns true if the element was added (new),
 * false if it already existed — no separate lookup needed.
 */
object HashSetFirstDuplicate {
  def solve(data: Array[Int]): Int = {
    val seen = mutable.HashSet[Int]()
    var i = 0
    while (i < data.length) {
      if (!seen.add(data(i))) return data(i)
      i += 1
    }
    -1
  }
}

/**
 * Implementation B: Using mutable.TreeSet (sorted set / red-black tree).
 *
 * Same algorithm as A, but the underlying container maintains sorted
 * order. Insert and lookup are O(log n) instead of O(1) amortized,
 * making overall time O(n log n). The tradeoff: iteration yields
 * elements in sorted order, and worst-case performance is more
 * predictable (no hash collisions).
 *
 * mutable.TreeSet.add has the same boolean-return interface as HashSet,
 * so the code is nearly identical.
 */
object TreeSetFirstDuplicate {
  def solve(data: Array[Int]): Int = {
    val seen = mutable.TreeSet[Int]()
    var i = 0
    while (i < data.length) {
      if (!seen.add(data(i))) return data(i)
      i += 1
    }
    -1
  }
}

File: rust/src/first_duplicate.rs
use std::collections::HashSet;
use std::collections::BTreeSet;

/// Implementation A: Using HashSet (hash set).
///
/// Iterate through the stream and insert each value into the hash set.
/// If insert returns false (the value was already present), we've found
/// the first duplicate. O(n) expected time, O(n) space.
///
/// HashSet::insert returns true if the value was added (new), false if
/// it already existed — no separate lookup needed.
pub fn first_duplicate_hash_set(data: &[i32]) -> i32 {
    let mut seen = HashSet::new();
    for &val in data {
        if !seen.insert(val) {
            return val;
        }
    }
    -1
}

/// Implementation B: Using BTreeSet (sorted set / B-tree).
///
/// Same algorithm as A, but the underlying container maintains sorted
/// order. Insert and lookup are O(log n) instead of O(1) amortized,
/// making overall time O(n log n). The tradeoff: iteration yields
/// elements in sorted order, and worst-case performance is more
/// predictable (no hash collisions).
///
/// BTreeSet::insert has the same boolean-return interface as HashSet,
/// so the code is nearly identical.
pub fn first_duplicate_sorted_set(data: &[i32]) -> i32 {
    let mut seen = BTreeSet::new();
    for &val in data {
        if !seen.insert(val) {
            return val;
        }
    }
    -1
}
